app/assets/javascripts/discourse/app/mixins/composer-upload-uppy.js

```diff
@@ -0,0 +1,468 @@
+          displayErrorForUpload(response, this.siteSettings, file.name);
+          this.appEvents.trigger("composer:upload-error", file);
+        }
+      });
+    });
+
+    this._uppyInstance.on("complete", () => {
+      this.appEvents.trigger("composer:all-uploads-complete");
+      this._reset();
+    });
+
+    this._uppyInstance.on("cancel-all", () => {
+      // uppyInstance.reset() also fires cancel-all, so we want to
+      // only do the manual cancelling work if the user clicked cancel
+      if (this.userCancelled) {
+        Object.values(this.placeholders).forEach((data) => {
+          this.appEvents.trigger(
+            "composer:replace-text",
+            data.uploadPlaceholder,
+            ""
+          );
+        });
+
+        this.set("userCancelled", false);
+        this._reset();
+
+        next(() => {
+          this.appEvents.trigger("composer:uploads-cancelled");
+        });
+      }
+    });
+
+    this._setupPreprocessing();
+  },
+
+  _setupPreprocessing() {
+    Object.keys(this.uploadProcessorActions).forEach((action) => {
+      switch (action) {
+        case "optimizeJPEG":
+          this._uppyInstance.use(UppyMediaOptimization, {
+            optimizeFn: this.uploadProcessorActions[action],
+            runParallel: !this.site.isMobileDevice,
+          });
+          this._trackPreProcessorStatus(UppyMediaOptimization);
+          break;
+      }
+    });
+
+    this._uppyInstance.on("preprocess-progress", (pluginClass, file) => {
+      this._preProcessorStatus[pluginClass].activeProcessing++;
+      let placeholderData = this.placeholders[file.id];
+      placeholderData.processingPlaceholder = `[${I18n.t(
+        "processing_filename",
+        {
+          filename: file.name,
+        }
+      )}]()\n`;
+
+      this.appEvents.trigger(
+        "composer:replace-text",
+        placeholderData.uploadPlaceholder,
+        placeholderData.processingPlaceholder
+      );
+    });
+
+    this._uppyInstance.on("preprocess-complete", (pluginClass, file) => {
+      run(() => {
+        let placeholderData = this.placeholders[file.id];
+        this.appEvents.trigger(
+          "composer:replace-text",
+          placeholderData.processingPlaceholder,
+          placeholderData.uploadPlaceholder
+        );
+        const preProcessorStatus = this._preProcessorStatus[pluginClass];
+        preProcessorStatus.activeProcessing--;
+        preProcessorStatus.completeProcessing++;
+
+        if (
+          preProcessorStatus.completeProcessing ===
+          preProcessorStatus.needProcessing
+        ) {
+          preProcessorStatus.allComplete = true;
+
+          if (this._allPreprocessorsComplete()) {
+            this.setProperties({
+              isProcessingUpload: false,
+              isCancellable: true,
+            });
+            this.appEvents.trigger("composer:uploads-preprocessing-complete");
+          }
+        }
+      });
+    });
+  },
+
+  _uploadFilenamePlaceholder(file) {
```

Comment:
This one is very similar to the regular composer upload mixin, however that one just stored one placeholder at a time on a shared property which was very hard to reason about. The _uploadPlaceholder method is now a pure function as well.